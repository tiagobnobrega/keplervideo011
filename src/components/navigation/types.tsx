/*
 * Copyright (c) 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */

import { DrawerScreenProps } from '@amzn/react-navigation__drawer/lib/typescript/src/types';
import { StackNavigationProp } from '@amzn/react-navigation__stack';
import { StackScreenProps } from '@amzn/react-navigation__stack/lib/typescript/src/types';
import { ImageSourcePropType, TouchableOpacity } from 'react-native';
import { TitleData } from '../../types/TitleData';

export enum Screens {
  DEFAULT_SCREEN = 'Home',
  APP_DRAWER = 'AppDrawer',
  HOME_SCREEN = 'Home',
  DETAILS_SCREEN = 'Details',
  PLAYER_SCREEN = 'Player',
  SETTINGS_SCREEN = 'Settings',
  SEARCH_SCREEN = 'Search',
  SEARCH_RESULTS_SCREEN = 'SearchResultsScreen',
  FEEDBACK_SCREEN = 'FeedBackScreen',
}

export enum DrawerType {
  PERMANENT = 'permanent',
  FRONT = 'front',
  SLIDE = 'slide',
  BACK = 'back',
}

export type AppStackParamList = {
  [Screens.APP_DRAWER]: undefined;
  [Screens.HOME_SCREEN]: undefined;
  [Screens.SEARCH_SCREEN]: undefined;
  [Screens.SETTINGS_SCREEN]: undefined;
  [Screens.FEEDBACK_SCREEN]: undefined;
  [Screens.SEARCH_RESULTS_SCREEN]: {
    searchKeyword: string;
  };
  [Screens.DETAILS_SCREEN]: {
    data: TitleData;
    sendDataOnBack: (id: number | string) => void;
  };
  [Screens.PLAYER_SCREEN]: {
    data: TitleData;
    sendDataOnBack: () => void;
  };
};

export type AppStackScreenProps<T extends keyof AppStackParamList> =
  StackScreenProps<AppStackParamList, T>;

export type HomeScreenNavigationProps = StackNavigationProp<
  AppStackParamList,
  Screens.HOME_SCREEN
>;

export type AppDrawerParamList = {
  [Screens.HOME_SCREEN]: undefined;
  [Screens.SETTINGS_SCREEN]: undefined;
  [Screens.SEARCH_SCREEN]: undefined;
};

export type AppDrawerScreenProps<T extends keyof AppDrawerParamList> =
  DrawerScreenProps<AppDrawerParamList, T>;

declare global {
  namespace ReactNavigation {
    interface RootParamList extends AppStackParamList {}
  }
}

export interface ButtonConfig {
  onPress: () => void;
  image: ImageSourcePropType;
  label: string;
  ref?: React.RefObject<TouchableOpacity>;
  testID: string;
}
