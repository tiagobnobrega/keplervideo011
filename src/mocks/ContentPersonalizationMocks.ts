import {
  ContentEntitlementBuilder,
  ContentIdBuilder,
  ContentIdNamespaces,
  ContentInteractionBuilder,
  ContentInteractionType,
  CustomerListEntryBuilder,
  CustomerListType,
  EntitlementType,
  IContentEntitlement,
  IContentId,
  IContentInteraction,
  ICustomerListEntry,
  IPlaybackEvent,
  IProfileId,
  PlaybackEventBuilder,
  PlaybackState,
  ProfileIdBuilder,
  ProfileIdNamespaces,
} from '@amzn/kepler-content-personalization';
import {
  ChannelDescriptorBuilder,
  IChannelDescriptor,
} from '@amzn/kepler-epg-provider';
import { VideoPlayer } from '@amzn/react-native-w3cmedia';

export function getMockContentID(id: string, namespace: string): IContentId {
  return new ContentIdBuilder().id(id).idNamespace(namespace).build();
}

export function getMockContentEntitlement(): IContentEntitlement {
  console.info('k_content_per: Building ContentEntitlement object in App');
  return new ContentEntitlementBuilder()
    .acquisitionTimestamp(new Date())
    .contentId(getMockContentID('testId', ContentIdNamespaces.NAMESPACE_CDF_ID))
    .entitlementType(EntitlementType.PURCHASE)
    .expirationTimestamp(new Date())
    .build();
}

export function getMockContentInteraction(
  interactionType: ContentInteractionType,
  content: IContentId,
): IContentInteraction {
  console.info('k_content_per: Building ContentInteraction object in App');
  return new ContentInteractionBuilder()
    .contentId(content)
    .interactionTimestamp(new Date())
    .contentInteractionType(interactionType)
    .profileId(getMockProfileId())
    .build();
}

export function getMockCustomerListEntry(
  listType: CustomerListType,
  content: IContentId,
): ICustomerListEntry {
  console.info('k_content_per: Building CustomerListEntry object in App');
  return new CustomerListEntryBuilder()
    .contentId(
      content
        ? content
        : getMockContentID('testId', ContentIdNamespaces.NAMESPACE_CDF_ID),
    )
    .addedTimestamp(new Date())
    .build();
}

export function getMockProfileId(): IProfileId {
  return new ProfileIdBuilder()
    .id('testProfile')
    .idNamespace(ProfileIdNamespaces.NAMESPACE_APP_INTERNAL)
    .build();
}

function getMockChannelDescriptor(): IChannelDescriptor {
  return new ChannelDescriptorBuilder()
    .identifier('UniqueChannelIdentifier')
    .majorNumber(1)
    .minorNumber(0)
    .build();
}

export function getDefaultMockPlaybackEvent(): IPlaybackEvent {
  console.info('k_content_per: Building PlaybackEvent object in App');
  return new PlaybackEventBuilder()
    .channelDescriptor(getMockChannelDescriptor())
    .contentId(getMockContentID('testId', ContentIdNamespaces.NAMESPACE_CDF_ID))
    .creditsPositionMs(5000)
    .durationMs(50000)
    .playbackPositionMs(100)
    .playbackState(PlaybackState.PAUSED)
    .profileId(getMockProfileId())
    .eventTimestamp(new Date())
    .buildActiveEvent();
}

export function getMockPlaybackEventForVideo(
  videoRef: React.MutableRefObject<VideoPlayer | null>,
  title: string,
  playbackState: PlaybackState,
): IPlaybackEvent {
  if (!videoRef.current) {
    throw new Error('k_content_per: Video player is not initialized');
  }
  return (
    new PlaybackEventBuilder()
      .playbackState(playbackState)
      .profileId(getMockProfileId())
      .contentId(
        getMockContentID('title', ContentIdNamespaces.NAMESPACE_CDF_ID),
      )
      // Multiplying it by 1000 since the duration retrieved from the player is in seconds
      .durationMs(videoRef.current.duration * 1000)
      // Multiplying it by 1000 since the currentTime retrieved from the player is in seconds
      .playbackPositionMs(videoRef.current.currentTime * 1000)
      .buildActiveEvent()
  );
}
