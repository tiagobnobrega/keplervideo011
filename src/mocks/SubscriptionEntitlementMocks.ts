import {
  ISubscriptionEntitlementEntry,
  SubscriptionEntitlementEntryBuilder,
} from '@amzn/kepler-subscription-entitlement';

export function getMockSubscriptionEntitlementEntry(): ISubscriptionEntitlementEntry {
  return new SubscriptionEntitlementEntryBuilder()
    .acquisitionTimestamp(new Date())
    .expirationTimestamp(new Date())
    .subscriptionId('testSubscriptionId')
    .build();
}
