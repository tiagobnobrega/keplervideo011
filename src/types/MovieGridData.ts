import { TitleData } from './TitleData';

export type MovieGridData = {
  heading: string;
  testID: string;
  data: () => TitleData[];
};
