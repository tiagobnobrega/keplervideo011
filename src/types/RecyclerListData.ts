import { StyleProp, ViewStyle } from 'react-native';
import { RecyclerListViewProps } from 'recyclerlistview';

export interface CustomRecyclerListViewProps extends RecyclerListViewProps {
  contentContainerStyle?: StyleProp<ViewStyle>;
  initialNumToRender: number | undefined;
}
