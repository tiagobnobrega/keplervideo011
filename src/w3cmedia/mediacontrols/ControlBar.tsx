import { VideoPlayer } from '@amzn/react-native-w3cmedia';
import React from 'react';
import { StyleSheet, View } from 'react-native';
import CaptionButton from './CaptionButton';

export interface ControlBarProps {
  captions: () => void;
  videoRef: React.MutableRefObject<VideoPlayer | null>;
  captionMenuVisibility: boolean;
  playerControlType?: string;
  isCaptionButtonFocused: boolean | null;
}
export const ControlBar = ({
  captions,
  videoRef,
  captionMenuVisibility,
  playerControlType,
  isCaptionButtonFocused,
}: ControlBarProps) => {
  return (
    <View style={styles.controlBar} testID="control-bar">
      <CaptionButton
        playerControlType={playerControlType}
        video={videoRef.current as VideoPlayer}
        onPress={captions}
        testID={'video-player-caption-btn'}
        captionMenuVisibility={captionMenuVisibility}
        isCaptionButtonFocused={isCaptionButtonFocused}
      />
    </View>
  );
};

export const styles = StyleSheet.create({
  controlBar: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
    alignItems: 'center',
  },
});
