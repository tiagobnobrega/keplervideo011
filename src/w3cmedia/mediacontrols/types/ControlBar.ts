import { VideoPlayer } from '@amzn/react-native-w3cmedia';

export enum PlayerControlType {
  PLAY = 'play',
  PAUSE = 'pause',
  PLAYPAUSE = 'playpause',
  SKIPBACKWARD = 'skip_backward',
  SKIPFORWARD = 'skip_forward',
}
export interface ControlBarButtonProps {
  videoRef: React.MutableRefObject<VideoPlayer | null>;
  onBlur?: () => void;
  playerControlType?: string;
}
