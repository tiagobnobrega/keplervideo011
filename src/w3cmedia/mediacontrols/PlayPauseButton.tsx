import {
  ContentPersonalizationServer,
  PlaybackState,
} from '@amzn/kepler-content-personalization';
import React, { useEffect, useState } from 'react';
import { TouchableOpacity } from 'react-native';
import { isContentPersonalizationEnabled } from '../../config/AppConfig';
import { getMockPlaybackEventForVideo } from '../../mocks/ContentPersonalizationMocks';
import PlayerButton from './PlayerButton';
import { ControlBarButtonProps } from './types/ControlBar';

const PlayPauseButton = React.forwardRef(
  (
    { videoRef, onBlur }: ControlBarButtonProps,
    ref: React.ForwardedRef<TouchableOpacity>,
  ) => {
    const initalPlayingState =
      !videoRef.current?.paused && !videoRef.current?.ended;
    const [playing, setPlaying] = useState<boolean>(initalPlayingState);

    useEffect(() => {
      addEventListeners();
      return () => {
        removeEventListeners();
      };
    });

    const onEndedUpdate = (): void => {
      if (videoRef.current) {
        //When the video ends, the play/pause icon does not change, and isPaused returns 0.
        //To handle that call pause API when the video ends.
        videoRef.current.pause();
        setPlaying(false);
      }
    };

    const addEventListeners = () => {
      videoRef.current?.addEventListener('play', onPlay);
      videoRef.current?.addEventListener('pause', onPause);
      videoRef.current?.addEventListener('ended', onEndedUpdate);
    };

    const removeEventListeners = () => {
      videoRef.current?.removeEventListener('play', onPlay);
      videoRef.current?.removeEventListener('pause', onPause);
      videoRef.current?.removeEventListener('ended', onEndedUpdate);
    };

    const onPause = () => {
      setPlaying(false);
    };

    const onPlay = () => {
      setPlaying(true);
    };

    const pause = () => {
      videoRef?.current?.pause();

      try {
        if (isContentPersonalizationEnabled()) {
          console.info('k_content_per: Creating playbackEvent object in pause');
          const playbackEvent = getMockPlaybackEventForVideo(
            videoRef,
            videoRef.current!.currentSrc,
            PlaybackState.PAUSED,
          );
          ContentPersonalizationServer.reportNewPlaybackEvent(playbackEvent);
          console.log(
            `k_content_per: Pause : Reporting new playback event :${JSON.stringify(
              playbackEvent,
            )}`,
          );
        }
      } catch (e) {
        console.error(`k_content_per: ${e}`);
      }
    };

    const play = () => {
      videoRef?.current?.play();
      try {
        if (isContentPersonalizationEnabled()) {
          console.info('k_content_per: Creating playbackEvent object in play');
          const playbackEvent = getMockPlaybackEventForVideo(
            videoRef,
            videoRef.current!.currentSrc,
            PlaybackState.PLAYING,
          );
          ContentPersonalizationServer.reportNewPlaybackEvent(playbackEvent);
          console.log(
            `k_content_per: Reporting new playback event : ${JSON.stringify(
              playbackEvent,
            )}`,
          );
        }
      } catch (e) {
        console.error(`k_content_per: ${e}`);
      }
    };
    return (
      <PlayerButton
        onPress={playing ? pause : play}
        icon={playing ? 'pause' : 'play-arrow'}
        size={70}
        ref={ref}
        onBlur={onBlur}
        testID="player-btn-play-pause"
      />
    );
  },
);

export default PlayPauseButton;
