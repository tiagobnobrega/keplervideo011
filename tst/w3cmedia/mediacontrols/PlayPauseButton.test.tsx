/*
 * Copyright (c) 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
import { VideoPlayer } from '@amzn/react-native-w3cmedia';
import { describe } from '@jest/globals';
import '@testing-library/jest-native/extend-expect';
import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import { consoleInfoSpy } from '../../../jest.setup';
import { getMockPlaybackEventForVideo } from '../../../src/mocks/ContentPersonalizationMocks';
import PlayPauseButton from '../../../src/w3cmedia/mediacontrols/PlayPauseButton';
import { ControlBarButtonProps } from '../../../src/w3cmedia/mediacontrols/types/ControlBar';

const pauseMock = jest.fn();
const playMock = jest.fn();
const videoRef: React.MutableRefObject<VideoPlayer | null> = {
  current: {
    paused: true,
    ended: true,
    play: playMock,
    pause: pauseMock,
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    currentSrc: 'PlayPause',
  } as unknown as VideoPlayer,
};
const renderCaptionButton = (props?: Partial<ControlBarButtonProps>) => {
  return <PlayPauseButton videoRef={videoRef} {...props} />;
};
describe('PlayPauseButton tests', () => {
  it('component renders correctly', () => {
    const component = render(renderCaptionButton());
    expect(component).toMatchSnapshot();
    const playerButton = component.getByTestId('player-btn-play-pause');
    fireEvent.press(playerButton);
    expect(playMock).toHaveBeenCalled();
  });
  it('component renders correctly with `paused` and `ended` negative', () => {
    const component = render(
      renderCaptionButton({
        videoRef: {
          current: {
            ...videoRef.current,
            paused: false,
            ended: false,
          } as unknown as VideoPlayer,
        },
      }),
    );
    expect(component).toMatchSnapshot();
    const playerButton = component.getByTestId('player-btn-play-pause');
    fireEvent.press(playerButton);
    expect(pauseMock).toHaveBeenCalled();
  });
});
const asMock = <T extends any = any, Y extends any[] = any[]>(
  obj: any,
): jest.Mock<T, Y> => obj as jest.Mock<T, Y>;
describe('PlayPauseButton event listeners', () => {
  test('Event `play`: Adds and Removes event listener correctly', () => {
    jest.spyOn(React, 'useRef').mockReturnValueOnce({ current: videoRef });
    const { unmount } = render(renderCaptionButton());
    let captureOnPlaying;
    expect(videoRef.current?.addEventListener).toHaveBeenCalledWith(
      'play',
      expect.any(Function),
    );
    captureOnPlaying = asMock(videoRef.current?.addEventListener).mock
      .calls[0][1];
    expect(typeof captureOnPlaying).toBe('function');
    captureOnPlaying();
    unmount();
    expect(videoRef.current?.removeEventListener).toHaveBeenCalledWith(
      'play',
      captureOnPlaying,
    );
  });
  test('Event `pause`: Adds and Removes event listener correctly', () => {
    jest.spyOn(React, 'useRef').mockReturnValueOnce({ current: videoRef });
    const { unmount } = render(renderCaptionButton());
    let captureOnPlaying;
    expect(videoRef.current?.addEventListener).toHaveBeenCalledWith(
      'pause',
      expect.any(Function),
    );
    captureOnPlaying = asMock(videoRef.current?.addEventListener).mock
      .calls[1][1];
    expect(typeof captureOnPlaying).toBe('function');
    captureOnPlaying();
    unmount();
    expect(videoRef.current?.removeEventListener).toHaveBeenCalledWith(
      'pause',
      captureOnPlaying,
    );
  });
  test('Event `ended`: Adds and Removes event listener correctly', () => {
    jest.spyOn(React, 'useRef').mockReturnValueOnce({ current: videoRef });
    const { unmount } = render(renderCaptionButton());
    let captureOnPlaying;
    expect(videoRef.current?.addEventListener).toHaveBeenCalledWith(
      'ended',
      expect.any(Function),
    );
    captureOnPlaying = asMock(videoRef.current?.addEventListener).mock
      .calls[2][1];
    expect(typeof captureOnPlaying).toBe('function');
    captureOnPlaying();
    unmount();
    expect(videoRef.current?.removeEventListener).toHaveBeenCalledWith(
      'ended',
      captureOnPlaying,
    );
  });
});

describe('PlayPauseButton onPress', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  it('function needs to be called on press with pause true', () => {
    const { getByTestId } = render(
      renderCaptionButton({
        videoRef: {
          current: {
            ...videoRef.current,
            paused: jest.fn(() => false),
            ended: jest.fn(() => false),
          } as unknown as VideoPlayer,
        },
      }),
    );
    playMock();
    fireEvent.press(getByTestId('player-btn-play-pause'));
    pauseMock();
    expect(videoRef.current?.pause).toHaveBeenCalled();
    expect(getMockPlaybackEventForVideo).toBeDefined();
  });
  it('function needs to be called on press', () => {
    const { getByTestId } = render(
      renderCaptionButton({
        videoRef: {
          current: {
            ...videoRef.current,
            paused: jest.fn(() => true),
            ended: jest.fn(() => true),
          } as unknown as VideoPlayer,
        },
      }),
    );
    fireEvent.press(getByTestId('player-btn-play-pause'));
    expect(videoRef.current?.play).toHaveBeenCalled();
    expect(consoleInfoSpy).toHaveBeenCalledWith(
      'k_content_per: Creating playbackEvent object in play',
    );
  });
});
