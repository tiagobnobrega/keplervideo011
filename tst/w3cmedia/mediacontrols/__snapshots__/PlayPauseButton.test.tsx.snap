// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayPauseButton tests component renders correctly 1`] = `
<View
  accessibilityState={
    Object {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    Object {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "opacity": 1,
      "padding": 25,
    }
  }
  testID="player-btn-play-pause"
>
  <Text />
</View>
`;

exports[`PlayPauseButton tests component renders correctly with \`paused\` and \`ended\` negative 1`] = `
<View
  accessibilityState={
    Object {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    Object {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "opacity": 1,
      "padding": 25,
    }
  }
  testID="player-btn-play-pause"
>
  <Text />
</View>
`;
